File: src/python/problem_011.py
import os
import operator
from functools import reduce

def load_input(filename):
    with open(filename) as fr:
        content = [list(map(int, line.strip().split(' '))) for line in fr.readlines()]
    return content

def compute_highest(values, k):
    return max([reduce(operator.mul, line[i:i+k]) for line in values
        for i in range(len(line) - k)])

# expects NxN matrix
def euler_011(values):
    k = 4
    size = len(values)

    # compute highest horizontal
    h1 = compute_highest(values, k)
    
    # compute highest vertical
    values2 = [[val[row] for val in values] for row, _ in enumerate(values)]
    h2 = compute_highest(values2, k)

    # compute highest diagonal (top left - bottom-right)
    values3 = [[1] * size for _ in range(size*2)]
    for r, row in enumerate(values):
        for c, n in enumerate(row):
            values3[size-r+c][r] = n
    h3 = compute_highest(values3, k)

    # compute highest diagonal (top right - bottom left)
    values4 = [[1] * size for _ in range(size*2)]
    for r, row in enumerate(values):
        for c, n in enumerate(row):
            values4[r+c][r] = n
    h4 = compute_highest(values4, k)

    return max([h1, h2, h3, h4])

if __name__ == '__main__':
    input_filename = os.path.join(os.path.dirname(__file__), 'input/problem_011.txt')
    input_value = load_input(input_filename)
    print(euler_011(input_value))

File: src/go/problem_011.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var (
	filepath = flag.String("input", "input/problem_011.txt", "file containing the input value")
)

func parseInt(n string) int {
	value, err := strconv.Atoi(n)
	if err != nil {
		panic(err)
	}
	return value
}

// get list of numbers from a line read from the file
func parseNumbers(line string) []int {
	entries := strings.Split(strings.Trim(line, " \n"), " ")
	nums := make([]int, len(entries))
	for i, e := range entries {
		nums[i] = parseInt(e)
	}
	return nums
}

func readInput(filename string) [][]int {
	file, err := os.Open(filename)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var lines [][]int
	for scanner := bufio.NewScanner(file); scanner.Scan(); {
		lines = append(lines, parseNumbers(scanner.Text()))
	}
	return lines
}

// find greatest product for a sequence of k values
// expects NxN input
func greatestProduct(values [][]int, k int) int {
	partialMax := 0

	// horizontal
	for i := 0; i < len(values); i++ {
		for j := 0; j < len(values[i]) - k; j++ {
			partialProduct := 1
			for offset := 0; offset < k; offset++ {
				partialProduct *= values[i][j + offset]
			}
			if partialProduct > partialMax {
				partialMax = partialProduct
			}
		}
	}

	// vertical
	for i := 0; i < len(values) - k; i++ {
		for j := 0; j < len(values[i]); j++ {
			partialProduct := 1
			for offset := 0; offset < k; offset++ {
				partialProduct *= values[i + offset][j]
			}
			if partialProduct > partialMax {
				partialMax = partialProduct
			}
		}
	}

	// diagonal 1
	for i := 0; i < len(values) - k; i++ {
		for j := 0; j < len(values[i]) - k; j++ {
			partialProduct := 1
			for offset := 0; offset < k; offset++ {
				partialProduct *= values[i + offset][j + offset]
			}
			if partialProduct > partialMax {
				partialMax = partialProduct
			}
		}
	}

	// diagonal 2
	for i := 0; i < len(values) - k; i++ {
		for j := 0; j < len(values[i]) - k; j++ {
			partialProduct := 1
			for offset := 0; offset < k; offset++ {
				partialProduct *= values[i + offset][j + k - offset]
			}
			if partialProduct > partialMax {
				partialMax = partialProduct
			}
		}
	}

	return partialMax
}

func main() {
	flag.Parse()

	inputValues := readInput(*filepath)

	fmt.Println(greatestProduct(inputValues, 4))
}
